### ISSUE TITLE
Support automatic generation of convenience builder methods

### ISSUE THREAD

**Issue by rolfw** [immutables/immutables#274]
A common feature of builders is that convenience methods are added that allow creating instances of the target type with less code. For example:

```java
@Value.Immutable
public interface Supplier {

    CompanyName name();  // -> a single builder method name(CompanyName value) will be generated. An overload name(String value) would be useful.

    @Value.Immutable(builder = false)
    interface CompanyName {    // -> ImmutableCompanyName with single factory method #of(String value) will be generated

        @Value.Parameter
        String value();
    }
}
```

To use this builder you often need an instance of `CompanyName`: `builder.name(ImmutableCompanyName.of("SomeCompany"))`

I would find it very useful if there would be a style option to generate convenience overloads for each factory method that exist on a property, so that a builder would be generated with an API like this: 

```java
public interface Builder {

    Builder name(CompanyName name);
    Builder name(String name);  // -> exists as a factory method on ImmutableCompanyName
}
```

The example here is a simple scalar value, but the benefit increases for composite value types that have a few more properties/parameters. I imagine this would be doable because factory methods have a distinct signature that are static and always return the type of their enclosing class. 

The parameter could be something like `@Value.Style(builderMethodOverloads = "*")` to generate overloads for all factory method of each property type (if any), or `@Value.Style(builderMethodOverloads = "of")` to generate only factory methods named `of(...)`. More complex criteria would probably need an annotation on the property itself.

**Comment by elucash** [comment]
Then, I guess, you'll enjoy using `Style.deepImmutablesDetection` attribute

**Comment by elucash** [comment]
Just found some example https://github.com/immutables/immutables/blob/master/value-fixture/src/org/immutables/fixture/deep/Canvas.java

**Comment by rolfw** [comment]
Yes, that's exactly what I was looking for. Thank you!

**Comment by rolfw** [comment]
I didn't immediately get it to work, by the way. Going through a process of elimination, I discovered that the reason it didn't work was because I was using the `copy = false` parameter on the value type. When I removed that it works.

The implementation doesn't actually use the copy methods though, so this seems like unexpected behavior. Is this a bug perhaps?

**Comment by rolfw** [comment]
I would also suggest to improve the Javadoc of the `Style.deepImmutablesDetection` attribute; When reading the javadoc it wasn't obvious to me what the effect would be. Also, when trying to solve the issue I just mentioned, I found myself looking for documentation on where to apply this style attribute (on the value type, on the enclosing type, on the declaring type, etc.).

Perhaps a simplified version of the canvas example would be good to include in the javadoc.

**Comment by rolfw** [comment]
Should `Style.deepImmutablesDetection` work on any generated type that contains other immutables? I'm now trying it on a type that includes a property of a type that is enclosed in another type and it doesn't generate the initializer methods corresponding to the factory methods.

An example usecase is the command pattern of DDD, like this command to create an instance of the supplier entity described above:

```java
@Value.Style(deepImmutablesDetection = true)interface AddSupplierCommand extends AddSupplierCommand {

    CompanyName name();
}
```

Here I would expect `ImmutableAddSupplierCommand.Builder` to have a method `nameOf(String value)`, but it doesn't.

**Comment by elucash** [comment]
Thank you for tripping on couple of landmines in this experimental functionality! ;)

* Bug to be fixed, code contains `if (valueType.isUseCopyConstructor())` it should have been `if (valueType.isUseConstructor())` in the first place, but some other refinements have to be made.
* You can read about styles at http://immutables.github.io/style.html, It does not go in details of each style (in favor of digging javadocs), but explains how to use styles in general.
* I'll improve javadoc a bit and will use code example there as you suggested, I'll also encourage anyone if have a proper wording in mind to contribute to the javadocs and documentation on the website ;)
* The last example works well for me if `CompanyName` itself has `@Value.Immutable` annotation. Deep immutables detection feature cannot work with `CompanyName` if it's was `Include`d because there are no reasonable way to find out if it's immutable or not without having immutable implementations. Any kind of brute force scanning will not work with incremental compilation and other peculiarities of the annotation processing.

**Comment by elucash** [comment]
Refinements released in 2.1.14, the code might need to be adjusted a bit (no `*Of` suffix)
Thank you!
